controllers/producer.go

```diff
@@ -0,0 +1,389 @@
+
+}
+
+func newPrometheusSourceConfig(env bootstrap.Environment) (metric.PrometheusSourceConfig, error) {
+	ps := env.Config.Metric.Prometheus
+	if ps == nil {
+		return metric.PrometheusSourceConfig{}, stderrors.New("failure create prometheus client, empty prometheus config")
+	}
+	promClient, err := prometheusApi.NewClient(prometheusApi.Config{
+		Address:      ps.Address,
+		RoundTripper: nil,
+	})
+	if err != nil {
+		return metric.PrometheusSourceConfig{}, err
+	}
+
+	return metric.PrometheusSourceConfig{
+		Api: prometheusV1.NewAPI(promClient),
+	}, nil
+}
+
+func accessLogHandler(logEntry []*data_accesslog.HTTPAccessLogEntry, svcToIps map[string][]string, cacheLock *sync.RWMutex) (map[string]map[string]string, error) {
+	log := log.WithField("reporter", "accesslog convertor").WithField("function", "accessLogHandler")
+	result := make(map[string]map[string]string)
+
+	tmpResult := make(map[string]map[string]int)
+	for _, entry := range logEntry {
+		//tmpValue := make(map[string]int)
+
+		// fetch sourceEp
+		sourceIp, err := fetchSourceIp(entry, log)
+		if err != nil {
+			return nil, err
+		}
+		if sourceIp == "" {
+			continue
+		}
+
+		// fetch sourceSvcMeta
+		sourceSvc, err := spliceSourceSvc(sourceIp, svcToIps, cacheLock, log)
+		if err != nil {
+			return nil, err
+		}
+		if sourceSvc == "" {
+			continue
+		}
+
+		// fetch destinationSvcMeta
+		destinationSvc := spliceDestinationSvc(entry, log)
+		if destinationSvc == "" {
+			continue
+		}
+
+		// push result
+		if dstSvcMappings, ok := tmpResult[sourceSvc]; !ok {
+			tmpValue := make(map[string]int)
+			tmpValue[destinationSvc] = 1
+			tmpResult[sourceSvc] = tmpValue
+		} else {
+			dstSvcMappings[destinationSvc] += 1
+		}
+
+		log.Debugf("tmpResult[%s][%s]: %d", sourceSvc, destinationSvc, tmpResult[sourceSvc][destinationSvc])
+	}
+
+	for sourceSvc, dstSvcMappings := range tmpResult {
+		result[sourceSvc] = make(map[string]string)
+		for dstSvc, count := range dstSvcMappings {
+			result[sourceSvc][dstSvc] = strconv.Itoa(count)
+		}
+	}
+
+	return result, nil
+}
+
+func fetchSourceIp(entry *data_accesslog.HTTPAccessLogEntry, log *logrus.Entry) (string, error) {
+	if entry.CommonProperties.DownstreamRemoteAddress == nil {
+		log.Debugf("DownstreamRemoteAddress is nil, skip")
+		return "", nil
+	}
+	downstreamSock, ok := entry.CommonProperties.DownstreamRemoteAddress.Address.(*envoy_config_core.Address_SocketAddress)
+	if !ok {
+		return "", stderrors.New("wrong type of DownstreamRemoteAddress")
+	}
+	if downstreamSock == nil || downstreamSock.SocketAddress == nil {
+		return "", stderrors.New("downstream socket address is nil")
+	}
+	log.Debugf("SourceEp is: %s", downstreamSock.SocketAddress.Address)
+	return downstreamSock.SocketAddress.Address, nil
+}
+
+func spliceSourceSvc(sourceIp string, svcToIps map[string][]string, cacheLock *sync.RWMutex, log *logrus.Entry) (string, error) {
+
+	cacheLock.RLock()
+	defer cacheLock.RUnlock()
+
+	for svc, ips := range svcToIps {
```

Comment:
pretty slow

Reply:
change to ipToSvcCache, and add svcToIpsCache as auxiliary map